app/models/events/recordable.rb

```diff
@@ -3,5 +3,24 @@ module Events::Recordable
 
   included do
     has_many :events, class_name: "Events::#{name}Event", dependent: :destroy, inverse_of: model_name.param_key
+
+    def record_event!(tag, request: Current.request, **additional)
+      ip_address = request&.ip_address
+      geoip_info = ip_address&.geoip_info
+
+      if (user_agent = request&.user_agent.presence)
+        begin
+          user_agent_info = Gemcutter::UserAgentParser.call(user_agent)
+          additional[:user_agent_info] = user_agent_info
+        rescue Gemcutter::UserAgentParser::UnableToParse => e
+          Rails.error.report(e, context: { user_agent: }, handled: true)
+        end
+      end
+
+      event = events.create!(tag:, ip_address:, geoip_info:, additional:, trace_id: Datadog::Tracing.correlation.trace_id)
+      logger.info("Recorded event #{tag}", record: cache_key,
```

Comment:
This might be kind of noisy. Any concerns about that?

Reply:
No concerns, these are all events that are helpful to be able to search in our logs as well.